### PR TITLE
[RFC] Remove the old x86 backend

### PR DESCRIPTION
**Pull request by bjorn3**
The default has been switched to the new x64 backend a while ago. AFAIK nobody has had any problems with the new x64 backend that required switching back to the old x86 backend.

This PR removes the old x86 backend, and the cranelift-codegen-meta part of the encoding and legalization mechanism of the old backend. The cranelift-codegen parts are only removed where necessary to fix warnings.

Based on #3007

### CODE REVIEW

**Comment by bjorn3** [comment]
The default was changed only two and a half months ago (https://github.com/bytecodealliance/wasmtime/pull/2718). It felt much longer.

**Comment by cfallin** [comment]
Thanks so much for tackling this! It has been on my to-do list for a while, but at low priority relative to a bunch of other things, so I'm happy that you had time to pick it up :-)

I think we should probably write up a proper RFC in bytecodealliance/rfcs, just to ensure we have the right folks onboard and there are no remaining surprise uses of the backend. I'll draft something soon and then we can start discussion.

**Comment by cfallin** [comment]
@bjorn3 now that the RFC is merged, would you be interested in bringing this PR up-to-date? If not, no worries, someone else can pick up the effort, but please do feel free if you're up for it, and I'm happy to review.

**Comment by bjorn3** [comment]
I will try to update it tomorrow.

**Comment by bjorn3** [comment]
@cfallin Rebased

**Comment by cfallin** [comment]
@bjorn3 there's a [build failure](https://github.com/bytecodealliance/wasmtime/pull/3009/checks?check_run_id=3746800417) that seems to arise from some missing DSL bits -- perhaps cut just a bit too much?

(As an aside, I'm very happy to see the -40k LoC stat; that's a lot of code we won't have to maintain!)

**Comment by bjorn3** [comment]
I forgot to remove two tests it seems. They were testing parts I cut out.

**Comment by bjorn3** [comment]
> (As an aside, I'm very happy to see the -40k LoC stat; that's a lot of code we won't have to maintain!)

~16k of deletions is tests for the old x86 backend or the old backend infrastructure. ~24k deletions are for cranelift-codegen.

**Comment by bjorn3** [comment]
```
FAIL filetests/filetests/runtests/extend.clif: run

Caused by:
    0: Cranelift codegen error
    1: Unsupported feature: Isplit: Unsupported type: types::I64
```

How did this test work on AArch64 in the first place? `isplit` from i64 -> i32, i32 shouldn't have been implemented on any new backend at all.

**Comment by cfallin** [comment]
> ```
> FAIL filetests/filetests/runtests/extend.clif: run
> 
> Caused by:
>     0: Cranelift codegen error
>     1: Unsupported feature: Isplit: Unsupported type: types::I64
> ```
> 
> How did this test work on AArch64 in the first place? `isplit` from i64 -> i32, i32 shouldn't have been implemented on any new backend at all.
> 
> Should I remove it?

Not sure, but for [that particular test](https://github.com/bytecodealliance/wasmtime/pull/3009/files#diff-f10ba883b9d5b1ca6f1b38e41be9d2eff0543e0e13fc67e169b4e63f6eb9c0e7R11) and its twin below, we could probably rewrite it without the `isplit` (maybe this is what you mean by "remove it") -- just a single 64-bit comparison rather than a split/two 32-bit compares/boolean-and.

**Comment by bjorn3** [comment]
Rewrote it to remove the usage of isplit.

File: cranelift/codegen/src/isa/mod.rs [hunk 1]
```diff
@@ -120,57 +111,24 @@ macro_rules! isa_builder {
...
-        }
-        (Architecture::X86_64, BackendVariant::Legacy) => {
-            isa_builder!(x86, (feature = "x86"), triple)
-        }
-        (Architecture::X86_64, BackendVariant::MachInst) => {
-            isa_builder!(x64, (feature = "x86"), triple)
-        }
-        #[cfg(not(feature = "old-x86-backend"))]
-        (Architecture::X86_64, BackendVariant::Any) => {
+pub fn lookup_variant(triple: Triple) -> Result<Builder, LookupError> {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/mod.rs:116]: Can you rename this to just `lookup` (and remove the wrapper below)? This was added to accept the extra `variant` arg and so I don't think we need it any more.

**Review by cfallin — APPROVED** [review]
Major thanks for this -- it was a large amount of work!

I went back and forth a bit about removing `BackendVariant` -- the plumbing exists now, what if we need it -- but on balance, I think the right choice is what's in this PR now, namely, providing a single consistent API and going back to the notion that there is just one backend per architecture. If in the future we have a really good reason to again duplicate backends then we can add plumbing again.

Also, I didn't look at the code that has *not* been removed in detail to work out what else could be removed, but I think there's still a lot: for example, any mention of `Encoding`s should eventually go away. But we can much more easily do that once the bulk of the old backend and associated infrastructure are removed; multiple followup PRs are expected and reasonable here, I think.

Anyway, LGTM and thanks again!

File: cranelift/codegen/src/isa/mod.rs (hunk 1 above)
Comment (bjorn3) [reply cranelift/codegen/src/isa/mod.rs]: Done

Comment (cfallin) [reply cranelift/codegen/src/isa/mod.rs:116]: It looks like the latest force-push lost this update?

Comment (bjorn3) [reply cranelift/codegen/src/isa/mod.rs]: Right, forgot to commit it again before force pushing.

**Comment by bjorn3** [comment]
CI passes

**Comment by cfallin** [comment]
Everything looks good -- well, here goes something! The old backend served us well, and thanks to those who worked on it; onward...

**Comment by dvc94ch** [comment]
I guess x86/riscv support is canned for the moment? as in there aren't enough users for these targets?

**Comment by bjorn3** [comment]
x86 is not a priority AFAIK, but it would be nice to have. Tracking issue: https://github.com/bytecodealliance/wasmtime/issues/1980

Riscv support has never been usable AFAICT. I believe it got prototyped with the original creation of Cranelift, but has never seen any improvements afterwards. Tracking issue: https://github.com/bytecodealliance/wasmtime/issues/2217

Nobody is currently working on either target to the best of my knowledge, but I am sure PR's for either will be accepted. @cfallin estimated that it will be 2-3 months of fulltime work to get to a reasonable state in https://github.com/bytecodealliance/wasmtime/issues/2217#issuecomment-961197939. If you do want to implement a backend, be aware of https://github.com/bytecodealliance/rfcs/pull/15.

**Pull request merged**